### DISCUSSION TITLE
[Feature]: Translation on web

### DISCUSSION THREAD

**Discussion by MalteKiefer** [immich-app/immich#1688] [Feature Request]
### Feature detail

It would be good if you could customize the web interface with its texts so that you could add translations.

### Platform

Web

**Comment by waclaw66** [comment]
This is highly anticipated feature.

**Comment by linguowei** [comment]
+1

**Comment by insomnia417** [comment]
+1

**Comment by bt90** [comment]
It would make sense to translate the mobile app using a platform such as Weblate. This way, translations could be shared between mobile and web at a later stage.

**Reply by alextran1502** [reply]
We are actually looking into Weblate right now

**Reply by bt90** [reply]
The syncthing project moved its projects to weblate in 2023. We went with their hosted service which is free for open source projects.

Discussion: https://forum.syncthing.net/t/the-old-transifex-thread-was-translating-syncthing/550/76

**Comment by Havouza15** [comment]
I can help with Swedish when necessary

**Comment by hectorzin** [comment]
And myself in spanish.... any way to do it?

**Comment by wqtty** [comment]
I posted a feature #8139  and it was closed as duplicated with this one, so any progress?

**Reply by JW-CH** [reply]
You posted a feature request for a feature that has already been requested. That's why it was closed. See https://github.com/immich-app/immich/discussions/1688#discussioncomment-8136214

**Comment by justkowal** [comment]
I'd love to provide Polish translations

**Comment by MickLesk** [comment]
I can do the German translation 👍🏻

**Comment by maicss** [comment]
I can do the Simple Chinese translation.

**Comment by alexisjanvier** [comment]
I can handle French translations if needed.

**Comment by AngelaDMerkel** [comment]
I can do Dutch translation (lol)

**Comment by nutzlichsein** [comment]
I can do Indonesian translation... how can I contribute?

**Comment by Deedikjupijn** [comment]
> We are actually looking into Weblate right now

For anyone wondering, this has not happened yet and so the only way to contribute with the web translations is if you were to implement weblate into the website (if there is any topic on this update I would love to hear it). Otherwise you would maybe be able to create a temporary button where people are able to change the languages and we just update every page manually (not using weblate). 

Hope this helps and hope everyone has a nice day!

**Comment by dianso** [comment]
i can handle chinese translation

**Comment by Aelbahat1** [comment]
> ### Feature detail
> It would be good if you could customize the web interface with its texts so that you could add translations.
> 
> ### Platform
> Web

How can I contact you with regards to the sentry app. I really need your help getting the logs of this app. I need to know the source of configuration to my device police. And why do I always have 410 applications running by default with hidden permissions. I am literally for the past 3 years have been manipulated and tortured by this. I need the help of a developer like you. #SOS

**Reply by Aelbahat1** [reply]
In return I can help you with any Arabic translation for free.

**Comment by Ador-able** [comment]
Is there any progress on this feature?

**Reply by MickLesk** [reply]
For the App, @alextran1502 uses localizley. I have no idea whether this is still planned for the web. I hope so :-)

**Reply by Manic-87** [reply]
Is the usage of Weblate now out of question? I actually wanted to try to look into it and try to bring translations to the web, but if Weblate is out of question I can obviously save me some time (and probably some headaches).

**Reply by bo0tzz** [reply]
Weblate is still planned, we just haven't gotten around to it. Same goes for translation support on the web.

**Reply by Manic-87** [reply]
Alright. Then I will invest some time looking into it, but absolutely no promises anything will come out of it.

**Comment by Manic-87** [comment]
So I did invest some time on this and I currently could display quite a few texts in different languages and also make use of weblate.
Disregarding the weblate portion, I want to make sure I'm on the right track about translatability of the web.

1. I added svelte-i18n as a dependency as it looked good to me (mind you, I don't have any experience with svelte, so I can't properly asses if this is actually good). This would be the first question: Is the usage of svelte-i18n ok or are there any concerns?
2. I started with the "page.svelte" files, took the hardcoded texts, moved them into a "en-EN.json" file and used the svelte-i18n functions to get the texts back. Regarding the structure of the file would be the next important question. For Android it's a flat list. For the web I thought about using a hierarchical structure as it, in my opinion at least, improves later readability/extendability. 
This means my file currently would look like this for example:
```
page: {
  admin: {
    library_management: {
      error_create: "Unable to create library"
      ...
      ...
    },
    system_settings: {
      auth_settings_title: "..."
      auth_settings_subtitle: "..."
      ...
      ...
    }
  }
}
```

For components the top level would be "component". For stuff which is used more often like "Video" or "User" there would be a top level "common" and so on. In the svelte files you'd then use 
`$t(page.admin.user_management.error_create)` 
for example. For stuff with placeholders you'd use  
`$t(page.admin.user_management.error_create, {values: {placeholdername: placeholdervalue}})`.
With the placeholders you can also use plurals, so instead of having two rules strings for ""Matched 1 item" and "Matched 10 items" there would only be one "ICU plural messages" `Matched {count, plural, one {# item} other {# items}}`

If you want you can also take a look at the actual translation file I'm currently building [here](https://github.com/immich-app/immich/compare/main...Manic-87:immich:feature/weblate#diff-8a308a21c04db59e7c6c44b7651a012634358492758e68264b7cae27f24bb4b0). There you also find how I currently replaced the texts in the svelte files. The branch itself was/is just a testing ground and all relevant changes would be moved to another clean branch.

What do you think? Is this the correct way for you or should I go at it some other way?

**Reply by bo0tzz** [reply]
Exciting! Can you start a draft PR with what you have right now? That provides a better place for discussing all the details. It doesn't matter if the branch is a mess/not ready.

**Reply by Manic-87** [reply]
Done #9854

**Reply by MickLesk** [reply]
Looks good. I'd have to do some translations and build the local, but it looks good so far. Would be worth a big release.

**Comment by saltfish13** [comment]
Hi there, I have been translating a Unity-based software, they use CSV files to store the lines and the corresponding keys so that the devs can refer to those translated lines using their keys in the UIs across different localizations. That said CSV files were posted on google docs so community translators can help out.  But it's a bit complicated I kenw there are off the shelf solution to do community localization but this is a free approach lol, I am willing to help out with SC(simplify Chinese) and TC (Traditional Chinese) localizations.

**Comment by bo0tzz** [comment]
Weblate is now set up for web translations at https://hosted.weblate.org/projects/immich/immich/. Feel free to sign up and have your hand at some translations!

**Reply by MickLesk** [reply]
Perfect. I do it today for germany
